app/models/manageiq/providers/microsoft/infra_manager/powershell.rb

```diff
@@ -10,13 +10,17 @@ def execute_powershell(connection, script)
       def run_powershell_script(connection, script)
         log_header = "MIQ(#{self.class.name}.#{__method__})"
         script_string = IO.read(script)
+
         begin
-          results = connection.shell(:powershell).run(script_string)
-          log_dos_error_results(results)
-          results
+          shell = connection.shell(:powershell)
+          results = shell.run(script_string)
+          log_dos_error_results(results.stderr)
+          results.stdout
         rescue Errno::ECONNREFUSED => err
           $scvmm_log.error "MIQ(#{log_header} Unable to connect to SCVMM. #{err.message})"
           raise
+        ensure
+          shell.close if shell
```

Comment:
I think [@Fryguy's comment](https://github.com/ManageIQ/manageiq/pull/13637/files#r97880216) is referring to this here.  Specifically, this *could* look something like:

```
def run_powershell_script(script)
  script_string = IO.read(script)
  with_provider_connection(:service => :powershell) do |shell|
    results = shell.run(script_string)
    log_dos_error_results(results.stderr)
    results.stdout
  end
end
```

And, everything else could be pulled up to the `with_provider_connection` method.  I think I could even go for a `with_powershell_connection` that's specific to this provider.  That way, the error handling could be as specific as necessary.

Then, the three methods that rely on Powershell connections wouldn't have to all copy the logic of cleanup.